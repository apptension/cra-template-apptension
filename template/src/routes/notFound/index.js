import { hot } from 'react-hot-loader/root';
import { NotFoundComponent } from './notFound.component';

export const NotFound = hot(NotFoundComponent);
