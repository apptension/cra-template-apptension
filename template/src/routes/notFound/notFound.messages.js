/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

export default defineMessages({
  pageTitle: {
    id: 'notFound.pageTitle',
    defaultMessage: 'Not found',
  },
  title: {
    id: 'notFound.title',
    defaultMessage: 'Error: 404',
  },
});
