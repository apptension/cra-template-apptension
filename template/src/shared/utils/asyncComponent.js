import loadable from '@loadable/component';

export const asyncComponent = (asyncLoader, exportedAs) => {
  return loadable(async () => {
    const module = await asyncLoader();
    return exportedAs ? module[exportedAs] : module;
  });
};
