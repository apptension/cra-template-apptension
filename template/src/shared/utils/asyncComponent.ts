import loadable from '@loadable/component';

export const asyncComponent = (asyncLoader: Function, exportedAs?: string) => {
  return loadable(async () => {
    const module = await asyncLoader();
    return exportedAs ? module[exportedAs] : module;
  });
};
