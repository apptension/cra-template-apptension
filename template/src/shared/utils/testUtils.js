import React from 'react';
import PropTypes from 'prop-types';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { identity } from 'ramda';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import { Route, Router } from 'react-router';
import { IntlProvider } from 'react-intl';
import { DEFAULT_LOCALE, translationMessages } from '../../i18n';
import { store as fixturesStore } from '../../../fixtures/store';

export * from '@testing-library/react';

export const PLACEHOLDER_CONTENT = <span data-testid="content">content</span>;

export const spiedHistory = (route = '/') => {
  const history = createMemoryHistory({ initialEntries: [route] });
  const pushSpy = jest.spyOn(history, 'push');
  return {
    history,
    pushSpy,
  };
};

export const ProvidersWrapper = ({ children, context = {} }) => {
  const { router = {}, store = fixturesStore, messages } = context;
  const { url = `/${DEFAULT_LOCALE}`, routePath = '/:lang/', history } = router;

  const routerHistory = history ?? createMemoryHistory({ initialEntries: [url] });

  const intlProviderMockProps = {
    locale: DEFAULT_LOCALE,
    messages: messages ?? translationMessages[DEFAULT_LOCALE],
  };

  return (
    <Router history={routerHistory}>
      <HelmetProvider>
        <IntlProvider {...intlProviderMockProps}>
          <Provider store={createStore(identity, store)}>
            <Route path={routePath}>{children}</Route>
          </Provider>
        </IntlProvider>
      </HelmetProvider>
    </Router>
  );
};

ProvidersWrapper.propTypes = {
  context: PropTypes.shape({
    router: PropTypes.shape({
      url: PropTypes.string,
      routePath: PropTypes.string,
      history: PropTypes.object,
    }),
    store: PropTypes.object,
    messages: PropTypes.object,
  }),
};

export const makeContextRenderer = component => (props = {}, context = {}) =>
  render(component(props), {
    wrapper: ({ children }) => <ProvidersWrapper context={context}>{children}</ProvidersWrapper>,
  });

export const makePropsRenderer = component => (props = {}) => render(component(props));
