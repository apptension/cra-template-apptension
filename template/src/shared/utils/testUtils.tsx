import React, { ReactNode, ReactElement, FC } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { identity } from 'ramda';
import { Provider } from 'react-redux';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Route, Router } from 'react-router';
import { IntlProvider } from 'react-intl';
import produce from 'immer';

import { DEFAULT_LOCALE, translationMessages, MessagesObject } from '../../i18n';
import { store as fixturesStore } from '../../../fixtures/store';
import { ResponsiveThemeProvider } from '../components/responsiveThemeProvider';
import { theme as defaultTheme } from '../../theme/theme';
import { GlobalState } from '../../modules/reducers';
import { LOCALES_INITIAL_STATE } from '../../modules/locales';
import { STARTUP_INITIAL_STATE } from '../../modules/startup';
import { USERS_INITIAL_STATE } from '../../modules/users';
//<-- IMPORT MODULE STATE -->

const defaultGlobalState: GlobalState = {
  locales: LOCALES_INITIAL_STATE,
  startup: STARTUP_INITIAL_STATE,
  users: USERS_INITIAL_STATE,
  //<-- INJECT MODULE STATE -->
};

export const PLACEHOLDER_TEST_ID = 'content';
export const PLACEHOLDER_CONTENT = <span data-testid="content">content</span>;

export const spiedHistory = (route = '/') => {
  const history = createMemoryHistory({ initialEntries: [route] });
  const pushSpy = jest.spyOn(history, 'push');
  return {
    history,
    pushSpy,
  };
};

interface ContextData {
  router?: {
    url?: string;
    routePath?: string;
    history?: MemoryHistory;
  };
  store?: GlobalState;
  messages?: MessagesObject;
  theme?: object;
}

interface ProvidersWrapperProps {
  children: ReactNode;
  context: ContextData;
}

export const ProvidersWrapper: FC<ProvidersWrapperProps> = ({ children, context = {} }) => {
  const { router = {}, store = fixturesStore, messages, theme = defaultTheme } = context;
  const { url = `/${DEFAULT_LOCALE}`, routePath = '/:lang/', history } = router;

  const routerHistory: MemoryHistory = history ?? createMemoryHistory({ initialEntries: [url] });

  const intlProviderMockProps = {
    locale: DEFAULT_LOCALE,
    messages: messages ?? translationMessages[DEFAULT_LOCALE],
  };

  return (
    <Router history={routerHistory}>
      <ResponsiveThemeProvider theme={theme}>
        <HelmetProvider>
          <IntlProvider {...intlProviderMockProps}>
            <Provider store={createStore(identity, store)}>
              <Route path={routePath}>{children}</Route>
            </Provider>
          </IntlProvider>
        </HelmetProvider>
      </ResponsiveThemeProvider>
    </Router>
  );
};

export const makeContextRenderer = <T, _>(component: (props: T | {}) => ReactElement) => (
  props?: T,
  context?: ContextData
) =>
  render(component(props ?? {}), {
    wrapper: ({ children }) => <ProvidersWrapper context={context}>{children}</ProvidersWrapper>,
  });

export const makePropsRenderer = <T, _>(component: (props: T | {}) => ReactElement) => (props?: T) =>
  render(component(props ?? {}));

export const prepareState = (stateSetter: (draftState: GlobalState) => void) =>
  produce(defaultGlobalState, stateSetter);
