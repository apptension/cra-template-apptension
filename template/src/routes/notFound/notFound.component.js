import React from 'react';
import { Helmet } from 'react-helmet-async';
import { FormattedMessage, useIntl } from 'react-intl';

import messages from './notFound.messages';
import { Container } from './notFound.styles';
import { H1 } from '../../theme/typography';

export const NotFoundComponent = () => {
  const { formatMessage } = useIntl();

  return (
    <Container>
      <Helmet title={formatMessage(messages.pageTitle)} />

      <H1>
        <FormattedMessage {...messages.title} />
      </H1>
    </Container>
  );
};
