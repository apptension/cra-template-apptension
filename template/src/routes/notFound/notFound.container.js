import React from 'react';
import { useIntl } from 'react-intl';

import { NotFoundComponent } from './notFound.component';

export const NotFoundContainer = () => {
  const { formatMessage } = useIntl();

  const componentProps = {
    formatMessage,
  };

  return <NotFoundComponent {...componentProps} />;
};
