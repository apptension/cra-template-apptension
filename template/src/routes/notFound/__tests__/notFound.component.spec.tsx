import React from 'react';

import { makeContextRenderer } from '../../../shared/utils/testUtils';
import { NotFoundComponent as NotFound } from '../notFound.component';

describe('NotFound: Component', () => {
  const component = () => <NotFound />;
  const render = makeContextRenderer(component);

  it('should render correctly', () => {
    const { container } = render();
    expect(container).toMatchSnapshot();
  });
});
