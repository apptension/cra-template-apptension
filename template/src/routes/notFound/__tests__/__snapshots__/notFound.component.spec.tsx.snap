// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFound: Component should render correctly 1`] = `
.c0 {
  font-family: OpenSans,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 39px;
  line-height: 43px;
}

<div>
  <div
    class=""
  >
    <h1
      class="c0"
    >
      Error: 404
    </h1>
  </div>
</div>
`;
